### PR TITLE
Contact form, that connects to a database

### PR DESCRIPTION
**Pull request by AbakirH**
Summary: I made a form that submits your contact information to a database, and it pulls the information from the database every time the page is reloaded.

Contact.java: Java file that can be used to create info about a person in form as a contact object

NewContactServlet.java - Servlet to make a new contact object to be pushed into a database using datastore API

ListContactsServlet.java - Servlet to get all the contacts in the database and return them all to the front end to be displayed on the screen when needed

DeleteContactServlet.java - Servlet to delete any contact entity from the database when the user sees fit. 

CommentServlet.java - This Servlet should not be in this pull request but is used to display any comment that the user may submit in the comment section of the website

http://abakir-step-2020.ue.r.appspot.com

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/delete-contact")
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java:14]: Please add a comment outlining the purpose of this class.

File: portfolio/src/main/java/com/google/sps/data/Contact.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+package com.google.sps.data;
+
+public final class Contact {
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/data/Contact.java:5]: Please add a doc outlining the purpose of this class.

File: portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    Query query = new Query("Contact").addSort("name", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Contact> contacts = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String name = (String) entity.getProperty("name");
+      String email = (String) entity.getProperty("email");
+      long number = (long) entity.getProperty("number");
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java:41]: Curious. Any particular reason for "long" over "int"

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -88,3 +88,53 @@ function createListElement(text) {
...
+    const taskListElement = document.getElementById('contact-list');
+    console.log(contacts);
+    contacts.forEach((contact) => {
+      taskListElement.appendChild(createContactElement(contact));
+     
+    })
+  });
+}
+function createContactElement(contact) {
+  const contactElement = document.createElement('li');
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js]: It's a good practice to keep literals as constants at the top of the file.

File: portfolio/src/main/java/com/google/sps/data/Contact.java (hunk 2 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/data/Contact.java:5]: do you mean comment?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java (hunk 1 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java:14]: okay

File: portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java (hunk 3 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java:41]: All the examples, that they gave us used long, so I thought that I should use it instead. Although a phone number should be under the maximum capacity of an int variable

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: that does make sense but in the case of these variables are only needed within the function, I feel I should leave them here. Also, all the code examples did this as well.

File: portfolio/src/main/java/com/google/sps/data/Contact.java (hunk 2 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/data/Contact.java:5]: nit: It's a good practice to keep line length to 80 characters or less.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java (hunk 1 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java:14]: *This
Also, Same comment as above w.r.t. line length.

File: portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java (hunk 3 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java:41]: Ok.

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (nightrunner500) [reply portfolio/src/main/webapp/script.js]: Sure. But these values ('li', 'Name: ", "Email: " etc) are not variables, they're constant literals.

Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: Makes sense will try to fix it lol

File: portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java (hunk 1 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java:14]: Was this updated?

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (nightrunner500) [reply portfolio/src/main/webapp/script.js]: Was this updated?

Also, github won't allow me to leave a comment there, but line 81 uses the literal 5. Can you store this in a constant at the top of this file. possibly with a comment describing what it's for?

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 5]
```diff
@@ -26,25 +26,34 @@
...
-    return comment;
+    try{
+      if(request.getParameter("comment") == null){
+        throw new Exception("Wrong Parameter path"); 
+      }
+      String comment = request.getParameter("comment");
+      return comment;
+    }catch(Exception path){
+      String comment = "Please check your code, because the path to get the comment is wrong";
+      return comment;
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: You're mixing two techniques here: if, and try/catch.

Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: I fixed the code like we discussed

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: It should be updated now, with literals in the code now

File: portfolio/src/main/java/com/google/sps/data/Contact.java (hunk 2 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/data/Contact.java:5]: Apologies that should have been docstring.

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 6]
```diff
@@ -26,25 +26,31 @@
...
-    comments.add(commenttext);
+    String commentText = getComment(request);
+    comments.add(commentText);
     response.sendRedirect("/index.html");
   }
+
   private String getComment(HttpServletRequest request) {
-    String comment = request.getParameter("comment");
+    String comment = ""; 
+    try{
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: please add spaces to this try / catch:
try { 
  ...
} catch (exception) { 
  ...
}

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 7]
```diff
@@ -26,25 +26,31 @@
...
+      comment = request.getParameter("comment");
+      return comment;
+    }catch(NullPointerException error){
+      comment = "Please check your code, because the path to get the comment is null";
+
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: Remove this line

File: portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for listing tasks. */
+@WebServlet("/list-contacts")
+public class ListContactsServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
```
Comment (nightrunner500) [inline portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java]: You can remove this line.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -88,3 +96,65 @@ function createListElement(text) {
...
 
+function loadContacts(){
+  fetch('/list-contacts')
+  .then(response => response.json())
+  .then((contacts) => {
+    const taskListElement = document.getElementById(contactPath);
+    console.log(contacts);
+    contacts.forEach((contact) => {
+      taskListElement.appendChild(createContactElement(contact));
+
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js]: Please remove this blank line.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -88,3 +96,65 @@ function createListElement(text) {
...
+  .then(response => response.json())
+  .then((contacts) => {
+    const taskListElement = document.getElementById(contactPath);
+    console.log(contacts);
+    contacts.forEach((contact) => {
+      taskListElement.appendChild(createContactElement(contact));
+     
+    });
+  });
+}
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js:107]: Please add a space before this method definition.

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -88,3 +96,65 @@ function createListElement(text) {
...
+function createContactElement(contact) {
+  contactIds.push(contact.id);
+
+  const contactElement = document.createElement('li');
+  contactElement.className = contactLiteral;
+
+  const nameElement = document.createElement('span');
+  nameElement.innerText = nameLiteral + ": " + contact.name;
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js]: See my comments above about this naming convention.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -15,9 +15,16 @@
 /**
  * Adds a random greeting to the page.
  */
+const contactLiteral = "contact";
+const nameLiteral = "Name";
+const emailLiteral = "Email";
+const numberLiteral = "Number";
+const deleteLiteral = "Delete";
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js]: These are not the most descriptive, and as such not that useful to the reader. Try to think about how the literals will be used, or what they identify. 

https://medium.com/coding-skills/clean-code-101-meaningful-names-and-functions-bf450456d90c

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -88,3 +96,65 @@ function createListElement(text) {
...
+    contactElement.remove();
+  });
+  contactElement.appendChild(nameElement);
+  contactElement.appendChild(emailElement);
+  contactElement.appendChild(numberElement);
+  contactElement.appendChild(deleteButtonElement);
+  
+  return contactElement;
+}
+function deleteTask(contact) {
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js:138]: Please add a space above this line.

Comment (AbakirH) [reply portfolio/src/main/webapp/script.js:138]: Fixed this

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js:107]: Fixed this

File: portfolio/src/main/webapp/script.js (hunk 9 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: Fixed this

File: portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java (hunk 8 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/ListContactsServlet.java]: Fixed this

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java (hunk 7 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: Fixed this

File: portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java (hunk 6 above)
Comment (AbakirH) [reply portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java]: Fixed this

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: should hopefully be good now

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (AbakirH) [reply portfolio/src/main/webapp/script.js]: I tried my best to fix this

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -15,16 +15,22 @@
 /**
  * Adds a random greeting to the page.
  */
+const nameOfForm = "contact";
+const nameOfContact = "Name";
+const emailOfContact = "Email";
+const numberOfContact = "Number";
+const deleteButton = "Delete";
```
Comment (nightrunner500) [inline portfolio/src/main/webapp/script.js:22]: "deleteButtonText" would be more accurate, as that's actually what this is.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java (hunk 1 above)
Comment (nightrunner500) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteContactServlet.java:14]: Please address this.

**Review by nightrunner500 — APPROVED** [review]

**Pull request merged**